Reference to `and` is ambiguous

<!-- Please read our Code of Conduct: https://github.com/deephaven/deephaven-core/blob/main/CODE_OF_CONDUCT.md -->
<!-- Please search existing issues to avoid creating duplicates. -->

**Description**

Attempting to use the `and` operator in query strings results in an error saying `reference to and is ambiguous`.

**Steps to reproduce**

From edge images (Python, and likely Groovy as well), create some data, and attempt to use `and` on it in a query string:

```python
from deephaven import new_table
from deephaven.column import bool_col

source = new_table([
    bool_col("A", [True, True, True]),
    bool_col("B", [True, True, False]),
    bool_col("C", [True, False, True])
])

result = source.update(["IsOk = and(A, B, C)"])
```

**Expected results**

The `result` table to contain the `IsOK` column, which is of `Boolean` or `boolean` type, and is only true if all of `A`, `B`, and `C` are true in the row.

**Actual results**

The following error (full stack trace):

```
r-Scheduler-Serial-1 | i.d.s.s.SessionState      | Internal Error 'b562d1e4-84a3-48bf-bb91-e8bcbc0f9903' java.lang.RuntimeException: Error in Python interpreter:
Type: <class 'deephaven.dherror.DHError'>
Value: table update operation failed. : RuntimeError: java.lang.RuntimeException: Error compiling class io.deephaven.temp.c_4736ceada8451154de90f4be2fcb88febfd4a1598105d20dfd7d9f62da3dda1ev61_0.Formula:
Traceback (most recent call last):
  File "/opt/deephaven/venv/lib/python3.10/site-packages/deephaven/table.py", line 784, in update
    return Table(j_table=self.j_table.update(*formulas))
RuntimeError: java.lang.RuntimeException: Error compiling class io.deephaven.temp.c_4736ceada8451154de90f4be2fcb88febfd4a1598105d20dfd7d9f62da3dda1ev61_0.Formula:
/io/deephaven/temp/c_4736ceada8451154de90f4be2fcb88febfd4a1598105d20dfd7d9f62da3dda1ev61_0/Formula.java:171: error: reference to and is ambiguous
            return and(A, B, C);
                   ^
  both method and(boolean...) in Logic and method and(Boolean...) in Logic match
Note: /io/deephaven/temp/c_4736ceada8451154de90f4be2fcb88febfd4a1598105d20dfd7d9f62da3dda1ev61_0/Formula.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 error

	at io.deephaven.engine.context.QueryCompiler.maybeCreateClassHelper(QueryCompiler.java:708)
	at io.deephaven.engine.context.QueryCompiler.maybeCreateClass(QueryCompiler.java:676)
	at io.deephaven.engine.context.QueryCompiler.compileHelper(QueryCompiler.java:455)
	at io.deephaven.engine.context.QueryCompiler.compile(QueryCompiler.java:262)
	at io.deephaven.engine.context.QueryCompiler.compile(QueryCompiler.java:215)
	at io.deephaven.engine.table.impl.select.DhFormulaColumn.compileFormula(DhFormulaColumn.java:798)
	at io.deephaven.engine.table.impl.select.DhFormulaColumn.createFormulaFactory(DhFormulaColumn.java:761)
	at io.deephaven.engine.table.impl.select.DhFormulaColumn.initDef(DhFormulaColumn.java:219)
	at io.deephaven.engine.table.impl.select.SwitchColumn.initDef(SwitchColumn.java:67)
	at io.deephaven.engine.table.impl.select.analyzers.SelectAndViewAnalyzer.create(SelectAndViewAnalyzer.java:104)
	at io.deephaven.engine.table.impl.select.analyzers.SelectAndViewAnalyzer.create(SelectAndViewAnalyzer.java:56)
	at io.deephaven.engine.table.impl.QueryTable.lambda$selectOrUpdate$33(QueryTable.java:1216)
	at io.deephaven.engine.table.impl.perf.QueryPerformanceRecorder.withNugget(QueryPerformanceRecorder.java:518)
	at io.deephaven.engine.table.impl.QueryTable.lambda$selectOrUpdate$34(QueryTable.java:1198)
	at io.deephaven.engine.table.impl.QueryTable.memoizeResult(QueryTable.java:3112)
	at io.deephaven.engine.table.impl.QueryTable.selectOrUpdate(QueryTable.java:1197)
	at io.deephaven.engine.table.impl.QueryTable.update(QueryTable.java:1176)
	at io.deephaven.engine.table.impl.QueryTable.update(QueryTable.java:102)
	at io.deephaven.api.TableOperationsDefaults.update(TableOperationsDefaults.java:90)
	at org.jpy.PyLib.executeCode(Native Method)
	at org.jpy.PyObject.executeCode(PyObject.java:138)
	at io.deephaven.engine.util.PythonEvaluatorJpy.evalScript(PythonEvaluatorJpy.java:73)
	at io.deephaven.integrations.python.PythonDeephavenSession.lambda$evaluate$1(PythonDeephavenSession.java:183)
	at io.deephaven.util.locks.FunctionalLock.doLockedInterruptibly(FunctionalLock.java:50)
	at io.deephaven.integrations.python.PythonDeephavenSession.evaluate(PythonDeephavenSession.java:182)
	at io.deephaven.engine.util.AbstractScriptSession.lambda$evaluateScript$1(AbstractScriptSession.java:145)
	at io.deephaven.engine.context.ExecutionContext.lambda$apply$0(ExecutionContext.java:129)
	at io.deephaven.engine.context.ExecutionContext.apply(ExecutionContext.java:140)
	at io.deephaven.engine.context.ExecutionContext.apply(ExecutionContext.java:128)
	at io.deephaven.engine.util.AbstractScriptSession.evaluateScript(AbstractScriptSession.java:145)
	at io.deephaven.engine.util.DelegatingScriptSession.evaluateScript(DelegatingScriptSession.java:87)
	at io.deephaven.engine.util.ScriptSession.evaluateScript(ScriptSession.java:113)
	at io.deephaven.server.console.ConsoleServiceGrpcImpl.lambda$executeCommand$4(ConsoleServiceGrpcImpl.java:170)
	at io.deephaven.server.session.SessionState$ExportBuilder.lambda$submit$2(SessionState.java:1352)
	at io.deephaven.server.session.SessionState$ExportObject.doExport(SessionState.java:896)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:162)
	at org.jpy.PyLib.callAndReturnObject(Native Method)
	at org.jpy.PyObject.call(PyObject.java:449)
	at io.deephaven.server.console.python.DebuggingInitializer.lambda$createInitializer$0(DebuggingInitializer.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)


Line: 786
Namespace: update
File: /opt/deephaven/venv/lib/python3.10/site-packages/deephaven/table.py
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/deephaven/venv/lib/python3.10/site-packages/deephaven/table.py", line 786, in update

        at org.jpy.PyLib.executeCode(PyLib.java:-2)
        at org.jpy.PyObject.executeCode(PyObject.java:138)
        at io.deephaven.engine.util.PythonEvaluatorJpy.evalScript(PythonEvaluatorJpy.java:73)
        at io.deephaven.integrations.python.PythonDeephavenSession.lambda$evaluate$1(PythonDeephavenSession.java:183)
        at io.deephaven.util.locks.FunctionalLock.doLockedInterruptibly(FunctionalLock.java:50)
        at io.deephaven.integrations.python.PythonDeephavenSession.evaluate(PythonDeephavenSession.java:182)
        at io.deephaven.engine.util.AbstractScriptSession.lambda$evaluateScript$1(AbstractScriptSession.java:145)
        at io.deephaven.engine.context.ExecutionContext.lambda$apply$0(ExecutionContext.java:129)
        at io.deephaven.engine.context.ExecutionContext.apply(ExecutionContext.java:140)
        at io.deephaven.engine.context.ExecutionContext.apply(ExecutionContext.java:128)
        at io.deephaven.engine.util.AbstractScriptSession.evaluateScript(AbstractScriptSession.java:145)
        at io.deephaven.engine.util.DelegatingScriptSession.evaluateScript(DelegatingScriptSession.java:87)
        at io.deephaven.engine.util.ScriptSession.evaluateScript(ScriptSession.java:113)
        at io.deephaven.server.console.ConsoleServiceGrpcImpl.lambda$executeCommand$4(ConsoleServiceGrpcImpl.java:170)
        at io.deephaven.server.session.SessionState$ExportBuilder.lambda$submit$2(SessionState.java:1352)
        at io.deephaven.server.session.SessionState$ExportObject.doExport(SessionState.java:896)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:162)
        at org.jpy.PyLib.callAndReturnObject(PyLib.java:-2)
        at org.jpy.PyObject.call(PyObject.java:449)
        at io.deephaven.server.console.python.DebuggingInitializer.lambda$createInitializer$0(DebuggingInitializer.java:30)
        at java.lang.Thread.run(Thread.java:833)
```

**Additional details and attachments**

If applicable, add any additional screenshots, logs, or other attachments to help explain your problem.

**Versions**
 - Deephaven: edge (no earlier than Thu Mar 16, 2023)
 - OS: OS X
 - Browser: Chrome
 - Docker: 20.10.13

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java
@@ -3,6 +3,8 @@
  */
 package io.deephaven.engine.table.impl.lang;
 
+import com.github.javaparser.JavaParser;
+import com.github.javaparser.StaticJavaParser;
 import com.github.javaparser.ast.ArrayCreationLevel;
 import com.github.javaparser.ast.CompilationUnit;
 import com.github.javaparser.ast.ImportDeclaration;
@@ -1102,8 +1104,8 @@ public final class QueryLanguageParser extends GenericVisitorAdapter<Class<?>, Q
                 }
             }
 
-            if (varArgType.isPrimitive() && allExpressionTypesArePrimitive) {
-                // there are ambiguous oddities with primitive varargs, so if its primitive lets just box it ourselves
+            if ((TypeUtils.isBoxedType(varArgType) || varArgType.isPrimitive()) && allExpressionTypesArePrimitive) {
+                // method invocation is ambiguous when both boxed and primitive versions of the method exist
                 Expression[] temp = new Expression[nArgs];
                 Expression[] varArgExpressions = new Expression[nArgExpressions - nArgs + 1];
                 System.arraycopy(expressions, 0, temp, 0, temp.length - 1);
@@ -1111,10 +1113,15 @@ public final class QueryLanguageParser extends GenericVisitorAdapter<Class<?>, Q
                         varArgExpressions.length);
 
                 NodeList<ArrayCreationLevel> levels = new NodeList<>(new ArrayCreationLevel());
+                com.github.javaparser.ast.type.Type elementType;
+                if (varArgType.isPrimitive()) {
+                    elementType = new PrimitiveType(PrimitiveType.Primitive.valueOf(
+                            varArgType.getSimpleName().toUpperCase()));
+                } else {
+                    elementType = StaticJavaParser.parseClassOrInterfaceType(varArgType.getSimpleName());
+                }
                 temp[temp.length - 1] = new ArrayCreationExpr(
-                        new PrimitiveType(
-                                PrimitiveType.Primitive.valueOf(varArgType.getSimpleName().toUpperCase())),
-                        levels, new ArrayInitializerExpr(new NodeList<>(varArgExpressions)));
+                        elementType, levels, new ArrayInitializerExpr(new NodeList<>(varArgExpressions)));
 
                 expressions = temp;
             }

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java
@@ -3,7 +3,6 @@
  */
 package io.deephaven.engine.table.impl.lang;
 
-import com.github.javaparser.JavaParser;
 import com.github.javaparser.StaticJavaParser;
 import com.github.javaparser.ast.ArrayCreationLevel;
 import com.github.javaparser.ast.CompilationUnit;

--- a/engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableSelectUpdateTest.java
+++ b/engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableSelectUpdateTest.java
@@ -1079,4 +1079,16 @@ public class QueryTableSelectUpdateTest {
         Assert.assertEquals(4L, updated.getColumnSource("D").get(updated.getRowSet().get(1)));
         Assert.assertEquals(8L, updated.getColumnSource("D").getPrev(updated.getRowSet().copyPrev().get(2)));
     }
+
+    @Test
+    public void testRegressionGH3562() {
+        final Table src = TableTools.emptyTable(3).update("A = true", "B = ii != 1", "C = ii != 2");
+        final Table result = src.select("And = and(A, B, C)", "Or = or(A, B, C)");
+
+        final Table expected = TableTools.newTable(
+                TableTools.col("And", true, false, false),
+                TableTools.col("Or", true, true, true));
+
+        assertTableEquals(expected, result);
+    }
 }

